flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeSubpartitionReader.java

```diff
@@ -19,154 +19,180 @@
+    @GuardedBy("lock")
+    private Throwable failureCause;
+
     /** Sequence number of the next buffer to be sent to the consumer. */
     private int sequenceNumber;
 
     public SortMergeSubpartitionReader(
-            int subpartitionIndex,
-            int dataBufferBacklog,
-            int bufferSize,
             SortMergeResultPartition partition,
             BufferAvailabilityListener listener,
-            PartitionedFile partitionedFile)
-            throws IOException {
+            PartitionedFileReader fileReader) {
         this.partition = checkNotNull(partition);
         this.availabilityListener = checkNotNull(listener);
-        this.dataBufferBacklog = dataBufferBacklog;
-
-        // allocate two pieces of unmanaged segments for data reading
-        for (int i = 0; i < NUM_READ_BUFFERS; i++) {
-            this.readBuffers.add(
-                    MemorySegmentFactory.allocateUnpooledOffHeapMemory(bufferSize, null));
-        }
-
-        this.fileReader = new PartitionedFileReader(partitionedFile, subpartitionIndex);
-        try {
-            readBuffers();
-        } catch (Throwable throwable) {
-            // ensure that the file reader is closed when any exception occurs
-            IOUtils.closeQuietly(fileReader);
-            throw throwable;
-        }
+        this.fileReader = checkNotNull(fileReader);
     }
 
     @Nullable
     @Override
     public BufferAndBacklog getNextBuffer() {
-        checkState(!isReleased, "Reader is already released.");
+        synchronized (lock) {
+            Buffer buffer = buffersRead.poll();
+            if (buffer == null) {
+                return null;
+            }
+
+            if (buffer.isBuffer()) {
+                --dataBufferBacklog;
+            }
 
-        Buffer buffer = buffersRead.poll();
-        if (buffer == null) {
-            return null;
+            Buffer lookAhead = buffersRead.peek();
+            return BufferAndBacklog.fromBufferAndLookahead(
+                    buffer,
+                    lookAhead == null ? Buffer.DataType.NONE : lookAhead.getDataType(),
+                    dataBufferBacklog,
+                    sequenceNumber++);
         }
+    }
 
-        if (buffer.isBuffer()) {
-            --dataBufferBacklog;
+    private void addBuffer(Buffer buffer) {
+        boolean notifyAvailable;
+        synchronized (lock) {
+            if (isReleased) {
+                buffer.recycleBuffer();
+                return;
+            }
+
+            notifyAvailable = buffersRead.isEmpty();
+
+            buffersRead.add(buffer);
+            if (buffer.isBuffer()) {
+                ++dataBufferBacklog;
+            }
         }
 
-        final Buffer lookAhead = buffersRead.peek();
+        if (notifyAvailable) {
+            notifyDataAvailable();
+        }
+    }
+
+    boolean readBuffers(Queue<MemorySegment> buffers, BufferRecycler recycler) throws IOException {
```

Comment:
We may add a comment here on that this method is running in separate IO threads.